pkg/certify/external-testdriver/driver-call.go

```diff
@@ -0,0 +1,342 @@
+		}
+
+		return testsuites.GetStorageClass(provisioner, parameters, nil, ns, suffix)
+	}
+
+	items, err := f.LoadFromManifests(d.StorageClass.FromFile)
+	Expect(err).NotTo(HaveOccurred(), "load storage class from %s", d.StorageClass.FromFile)
+	Expect(len(items)).To(Equal(1), "exactly one item from %s", d.StorageClass.FromFile)
+
+	err = f.PatchItems(items...)
+	Expect(err).NotTo(HaveOccurred(), "patch items")
+
+	sc, ok := items[0].(*storagev1.StorageClass)
+	Expect(ok).To(BeTrue(), "storage class from %s", d.StorageClass.FromFile)
+	if fsType != "" {
+		if sc.Parameters == nil {
+			sc.Parameters = map[string]string{}
+		}
+		sc.Parameters["csi.storage.k8s.io/fstype"] = fsType
+	}
+	return sc
+}
+
+func (d driverDefinition) GetSnapshotClass(config *testsuites.PerTestConfig) *unstructured.Unstructured {
+	if !d.SnapshotClass.FromName {
+		framework.Skipf("Driver %q does not support snapshotting - skipping", d.DriverInfo.Name)
+	}
+
+	snapshotter := config.GetUniqueDriverName()
+	parameters := map[string]string{}
+	ns := config.Framework.Namespace.Name
+	suffix := snapshotter + "-vsc"
+
+	return testsuites.GetSnapshotClass(snapshotter, parameters, ns, suffix)
+}
+
+func (d *driverDefinition) GetClaimSize() string {
+	return d.ClaimSize
+}
+
+func (d *driverDefinition) PrepareTest(f *framework.Framework) (*testsuites.PerTestConfig, func()) {
+	config := &testsuites.PerTestConfig{
+		Driver:         d,
+		Prefix:         "external",
+		Framework:      f,
+		ClientNodeName: d.ClientNodeName,
+	}
+	return config, func() {}
+}
+
+//Example call: getCommand("nfs", createVolume)
+func execCommand(pluginName string, cmdName string) []byte {
+	setCurrentNameSpace(currentNameSpace)
+
+	fmt.Printf("Command = %s", ". ../../pkg/certify/external-testdriver/"+pluginName+" && "+cmdName)
+
+	cmd := exec.Command("bash", "-c", ". ../../pkg/certify/external-testdriver/"+pluginName+" && "+cmdName)
+	var out bytes.Buffer
+	cmd.Stdout = &out
+
+	err := cmd.Run()
+	if err != nil {
+		fmt.Printf("Unable to execute command: %s \n", cmdName)
+		fmt.Printf(err.Error())
+	} else {
+		fmt.Printf("Executing command %s \n %s", cmdName, out.String())
+		return out.Bytes()
+	}
+
+	setCurrentNameSpace("default")
+
+	return nil
+}
+
+func setCurrentNameSpace(namespace string) error {
+	currentContextCmd := exec.Command("bash", "-c", "kubectl config current-context")
+	currentContextOutput, err := currentContextCmd.CombinedOutput()
+
+	if err != nil {
+		fmt.Printf("Unable to get current context")
+		fmt.Printf(err.Error())
+		return err
+	}
+
+	cmd := exec.Command("bash", "-c", "kubectl config set-context "+strings.TrimSpace(string(currentContextOutput))+" --namespace="+namespace+"")
+	output, err := cmd.CombinedOutput()
+	fmt.Printf("Kubectl output: %s\n\n", string(output))
+
+	if err != nil {
+		fmt.Printf("Unable to execute Kubectl command to set namespace")
+		fmt.Printf(err.Error())
+		return err
+	}
+
+	return nil
+
+}
+
+func getTestDriverType() {
```

Comment:
should this return something?

Reply:
can just move that code to Set function (Where it is being called)

Reply:
yeah I would avoid global var like validTestDriver, I did not see it at first so I thought this function did nothing. the Set function can call checkBashFuncExists directly so it doesn't need the validTestDriver var. i guess preprovisionedVolumeTestDriver and preprovisionedPVTestDriver could be fields of the DriverCall struct and then you can try to remove the driverDefinition struct and just work with the DriverCall struct. DriverCall can implement utils.DriverDefinition. It's not clear the relationship between driverDefinition and DriverCall, I know driverDefinition is copied from the other thing, but I would try to use just DriverCall here

Reply:
Hmm, so this bash driver TestDriver is basically a superset of `driverDefinition` so that's why it is confusing in terms of the structs and naming and such.

In pkg/external, we have a TestDriver implementation called `driverDefinition` plus a parameter/flag `DriverDefParameter` for constructing a driverDefinition.

I think we should copy that style and in pkg/external-bash (renamed from pkg/external-testdriver) have a TestDriver implementation called `bashDriver` (that is very similar to `driverDefinition` EXCEPT for when it makes createvolume etc. calls) plus a parameter/flag `BashDriverParameter` for constructing a `bashDriver`. Does that make sense?

Reply:
just committed something some edits before reading this, will look at it again

Reply:
ya that makes sense, the naming of the interfaces I have are definitely confusing

Reply:
I made changes in my last commit, let me know if this is what you meant.